bazel_repositories_priv

```diff
@@ -0,0 +1,292 @@
+        stdout += line
+    p.stdout.close()
+    returncode = p.wait()
+
+    urls = re.findall(r"https?://(?:[a-zA-Z]|[0-9]|[-_@.&/+])+", stdout)
+    urls = list(dict.fromkeys(urls))
+
+    conn.send((returncode, urls,))
+    conn.close()
+
+
+def chown(pathname: str):
+    """Change the owner of the given file from root to the unprivileged user.
+    """
+    user_passwd = pwd.getpwnam(user)
+    log.debug(f"chown {pathname}")
+    os.chown(pathname, user_passwd.pw_uid, user_passwd.pw_gid)
+    for root, dirs, files in os.walk(pathname):
+        for d in dirs:
+            filename = os.path.join(root, d)
+            log.debug(f"chown {filename}")
+            os.chown(filename, user_passwd.pw_uid, user_passwd.pw_gid)
+        for f in files:
+            filename = os.path.join(root, f)
+            log.debug(f"chown {filename}")
+            os.chown(filename, user_passwd.pw_uid, user_passwd.pw_gid)
+
+
+def download(url: str, dst: str):
+    log.info(f"Downloading {url}")
+    urllib_request.urlretrieve(url, dst)
+    log.info("Download finished")
+
+
+def process_url(url: str, url_path: pathlib.Path, filename: str) -> str:
+    """Download and store the archive from the given URL.
+    """
+    # If two first chars are not alpha, it means that the tarball name is a
+    # version or a commit hash. In that case, append the project name for
+    # readability.
+    if not (filename[0].isalpha() and filename[1].isalpha()):
+        prefix = None
+        # URLs from mirror.bazel.build have path like i.e.:
+        # mirror.bazel.build/github.com/account/project
+        # so the project name is the 3rd element of the path (which does not
+        # include the domain).
+        if "mirror.bazel.build" in url:
+            prefix = url_path.parts[3]
+        # And we assume that the other tarballs are downloaded from git hosting
+        # services, where path usually looks like i.e.:
+        # github.com/account/project
+        # gitlab.com/account/project
+        # so the project name is the 2nd element of the path (which does not
+        # include the domain).
+        else:
+            prefix = url_path.parts[2]
+        filename = prefix + "-" + filename
+    download(url, filename)
+
+    log.info(f"Checking sha256 sum of {filename}")
+    sha256sum = hashlib.sha256()
+    with open(filename, "rb") as f:
+        while True:
+            buf = f.read(65536)
+            if not buf:
+                break
+            sha256sum.update(buf)
+    sha256sum_digest = sha256sum.hexdigest()
+
+    log.info(f"Storing {filename}")
+    os.makedirs(os.path.join(CACHEDIR, sha256sum_digest), exist_ok=True)
+    cache_filename = os.path.join(CACHEDIR, sha256sum_digest, "file")
+    shutil.move(filename, cache_filename)
+
+    return cache_filename
+
+
+def process_urls(urls: typing.List[str],
+                 exclude: typing.List[str]) -> typing.Tuple[
+                     typing.List[str], typing.List[str]]:
```

Comment:
This annotation's indentation feels weird, but maybe that's just me.  What does flake8 have to say about it?

Reply:
flake8 is fine with that. I was not sure how should I indent type hinting. I can try to do it in a different way which will be fine for flake8.

Reply:
I was hesitant to leave this comment, since I don't really have a better suggestion. :D